docs/design/proposals/late-initialization/late-initialization.md

```diff
@@ -0,0 +1,552 @@
+        }
+    ```
+
+3. **Provide 3 new hooks to AWS service teams to customize the code generation in step 4**
+    
+    These 3 hooks will be:
+    
+    a) **set_late_initialized_defaults_override**: This hook will provide the **complete implementation** for setting default values of optional spec fields which are not provided by the k8s user and defaulted by AWS service.
+    This hook is mutually exclusive with "pre_set_late_initialized_defaults" & "post_set_late_initialized_defaults" hooks.
+
+    b) **pre_set_late_initialized_defaults**:- This hook will provide the custom code to be inserted **before** setting server side defaults for missing optional spec fields.
+
+    c) **post_set_late_initialized_defaults**: This hook will provide the custom code to be inserted **after** setting server side defaults for missing optional spec fields.
+
+4. Add `AWSResourceManager.SetLateInitializedDefaults()` method to handle server side defaults for fields mentioned in LateInitializationConfig.
+
+* Proposed Pseudo Code
+    ```go
+    // manager.go.tpl
+    func (rm *resourceManager) SetLateInitializedDefaults(
+	ctx context.Context,
+    latest acktypes.AWSResource,
+    ) (acktypes.AWSResource, error) {
+        {{- if $hookCode := Hook .CRD "set_late_initialized_defaults_override" }}
+            {{ $hookCode }}
+        {{- else}}
+            {{- if $hookCode := Hook .CRD "pre_set_late_initialized_defaults" }}
+                {{ $hookCode }}
+            {{- end }}
+
+            // 1. Filter all the fields which have LateInitializationConfig.
+            // 2. Filter out the fields from step1 with nil value in latest object.
+            // 3. If number of fields after step2 are nonzero, calculate the delay
+            // for late initialization.
+            // 4. If the latest object has 'LateInitialized=False' condition,
+            // no need to delay the late initialization. Otherwise calculate
+            // net delay for late initialization.(why? see step 7).
+            // 5. For all the fields with DefaultImmidiately=True, delay is
+            // zero, for fields with DefaultImmidiately=False, delay is
+            // 'DelaySeconds', with 5 seconds as default DelaySeconds
+            // 6. Net delay is maximum delay from all the Fields.
+            // 7. If net delay is more than zero, add 'LateInitialized=False'
+            // condition to the resource status indicating that resource is still
+            // being late initalized and return with requeue delay as net delay.
+            // 8. Make the ReadOne call.
+            // 9. Filter out the fields from step2, which are not-nil in ReadOne output.
+            // 10. For all the fields in step 9, set the lateInitialized value.
+            // 11. Upsert resource condition 'LateInitialized=True'.
+
+            {{- if $hookCode := Hook .CRD "post_set_late_initialized_defaults" }}
+                {{ $hookCode }}
+            {{- end }}
```

Comment:
given where the proposed hook points are, let's call the two hooks:

`late_initialize_pre_read_one`, `late_initialize_post_read_one`